src/ddmd/lexer.d

```diff
@@ -253,6 +270,12 @@ class Lexer
         t.lineComment = null;
         while (1)
         {
+            // Return error token when end of buffer is reached unexpectedly.
+            if (p >= end) {
```

Comment:
Lexing performance is critical to compiler speed, and this adds a compare/branch for every character processed. This is why a comparison to the buffer end is not done, and a sentinal (0 termination) is done. The sentinel does not require any extra code in the main path.